FWCore/Framework/src/ScheduleBuilder.cc

```diff
@@ -0,0 +1,728 @@
+                                                      iActivityRegistry,
+                                                      &iPrealloc,
+                                                      iProcessConfiguration,
+                                                      end_path_name,
+                                                      iEndPathNames,
+                                                      conditionalTaskHelper,
+                                                      conditionalModules));
+      for (auto const& path : endpathNameAndModules_.back().second) {
+        modulesInPaths.insert(path.description_->moduleLabel());
+      }
+    }
+
+    allNeededModules_.reserve(modulesInPaths.size());
+    iModuleRegistry.forAllModuleHolders([&](auto const* holder) {
+      if (modulesInPaths.find(holder->moduleDescription().moduleLabel()) != modulesInPaths.end()) {
+        allNeededModules_.emplace_back(&holder->moduleDescription());
+      }
+    });
+
+    std::vector<std::string> modulesInConfig(ioProcessPSet.getParameter<std::vector<std::string>>("@all_modules"));
+    std::set<std::string> modulesInConfigSet(modulesInConfig.begin(), modulesInConfig.end());
+    std::vector<std::string> unusedLabels;
+    set_difference(modulesInConfigSet.begin(),
+                   modulesInConfigSet.end(),
+                   modulesInPaths.begin(),
+                   modulesInPaths.end(),
+                   back_inserter(unusedLabels));
+    std::set<std::string> unscheduledLabels;
+    std::vector<std::string> shouldBeUsedLabels;
+    if (!unusedLabels.empty()) {
+      for (auto const& label : unusedLabels) {
+        auto modType =
+            ioProcessPSet.getParameter<edm::ParameterSet>(label).getParameter<std::string>("@module_edm_type");
+        if (modType == kProducerType || modType == kFilterType) {
+          auto module = getModule(ioProcessPSet,
+                                  label,
+                                  iModuleRegistry,
+                                  ioProductRegistry,
+                                  iActivityRegistry,
+                                  &iPrealloc,
+                                  iProcessConfiguration);
+
+          unscheduledModules_.push_back(&module->moduleDescription());
+          unscheduledLabels.emplace(label);
+        } else {
+          shouldBeUsedLabels.push_back(label);
+        }
+      }
+      if (!shouldBeUsedLabels.empty()) {
+        std::ostringstream unusedStream;
+        unusedStream << "'" << shouldBeUsedLabels.front() << "'";
+        for (std::vector<std::string>::iterator itLabel = shouldBeUsedLabels.begin() + 1,
+                                                itLabelEnd = shouldBeUsedLabels.end();
+             itLabel != itLabelEnd;
+             ++itLabel) {
+          unusedStream << ",'" << *itLabel << "'";
+        }
+        LogInfo("path") << "The following module labels are not assigned to any path:\n" << unusedStream.str() << "\n";
```

Comment:
I'm tempted with the following even if it is strictly speaking outside of the scope of this PR
```suggestion
        LogInfo("path").log([&shouldBeUsedLabels](auto& l) {
          l << "The following module labels are not assigned to any path:\n";
          l << "'" << shouldBeUsedLabels.front() << "'";
          for (std::vector<std::string>::iterator itLabel = shouldBeUsedLabels.begin() + 1,
                                                  itLabelEnd = shouldBeUsedLabels.end();
               itLabel != itLabelEnd;
               ++itLabel) {
            l << ",'" << *itLabel << "'";
          }
        });
```